File: Java/CoreNative/src/main/java/com/breadwallet/corenative/crypto/BRCryptoStatus.java
/*
 * Created by Michael Carrara <michael.carrara@breadwallet.com> on 7/1/19.
 * Copyright (c) 2019 Breadwinner AG.  All right reserved.
*
 * See the LICENSE file at the project root for license information.
 * See the CONTRIBUTORS file at the project root for a list of contributors.
 */
package com.breadwallet.corenative.crypto;

import com.google.common.collect.ImmutableMap;

import static com.google.common.base.Preconditions.checkState;

public enum BRCryptoStatus {

    CRYPTO_SUCCESS {
        @Override
        public int toNative() {
            return 0;
        }
    },

    CRYPTO_ERROR_FAILED {
        @Override
        public int toNative() {
            return 1;
        }
    },

    // Reference access
    CRYPTO_ERROR_UNKNOWN_NODE {
        @Override
        public int toNative() {
            return 10000;
        }
    },
    CRYPTO_ERROR_UNKNOWN_TRANSFER {
        @Override
        public int toNative() {
            return 10001;
        }
    },
    CRYPTO_ERROR_UNKNOWN_ACCOUNT {
        @Override
        public int toNative() {
            return 10002;
        }
    },
    CRYPTO_ERROR_UNKNOWN_WALLET {
        @Override
        public int toNative() {
            return 10003;
        }
    },
    CRYPTO_ERROR_UNKNOWN_BLOCK {
        @Override
        public int toNative() {
            return 10004;
        }
    },
    CRYPTO_ERROR_UNKNOWN_LISTENER {
        @Override
        public int toNative() {
            return 10005;
        }
    },

    // Node
    CRYPTO_ERROR_NODE_NOT_CONNECTED {
        @Override
        public int toNative() {
            return 20000;
        }
    },

    // Transfer
    CRYPTO_ERROR_TRANSFER_HASH_MISMATCH {
        @Override
        public int toNative() {
            return 30000;
        }
    },
    CRYPTO_ERROR_TRANSFER_SUBMISSION {
        @Override
        public int toNative() {
            return 30001;
        }
    },

    // Numeric
    CRYPTO_ERROR_NUMERIC_PARSE {
        @Override
        public int toNative() {
            return 40000;
        }
    };

    private static final ImmutableMap<Integer, BRCryptoStatus> LOOKUP;

    static {
        ImmutableMap.Builder<Integer, BRCryptoStatus> b = ImmutableMap.builder();

        b.put(CRYPTO_SUCCESS.toNative(),                      CRYPTO_SUCCESS);
        b.put(CRYPTO_ERROR_FAILED.toNative(),                 CRYPTO_ERROR_FAILED);

        b.put(CRYPTO_ERROR_UNKNOWN_NODE.toNative(),           CRYPTO_ERROR_UNKNOWN_NODE);
        b.put(CRYPTO_ERROR_UNKNOWN_TRANSFER.toNative(),       CRYPTO_ERROR_UNKNOWN_TRANSFER);
        b.put(CRYPTO_ERROR_UNKNOWN_ACCOUNT.toNative(),        CRYPTO_ERROR_UNKNOWN_ACCOUNT);
        b.put(CRYPTO_ERROR_UNKNOWN_WALLET.toNative(),         CRYPTO_ERROR_UNKNOWN_WALLET);
        b.put(CRYPTO_ERROR_UNKNOWN_BLOCK.toNative(),          CRYPTO_ERROR_UNKNOWN_BLOCK);
        b.put(CRYPTO_ERROR_UNKNOWN_LISTENER.toNative(),       CRYPTO_ERROR_UNKNOWN_LISTENER);

        b.put(CRYPTO_ERROR_NODE_NOT_CONNECTED.toNative(),     CRYPTO_ERROR_NODE_NOT_CONNECTED);

        b.put(CRYPTO_ERROR_TRANSFER_HASH_MISMATCH.toNative(), CRYPTO_ERROR_TRANSFER_HASH_MISMATCH);
        b.put(CRYPTO_ERROR_TRANSFER_SUBMISSION.toNative(),    CRYPTO_ERROR_TRANSFER_SUBMISSION);

        b.put(CRYPTO_ERROR_NUMERIC_PARSE.toNative(),          CRYPTO_ERROR_NUMERIC_PARSE);

        LOOKUP = b.build();
    }

    public static BRCryptoStatus fromNative(int nativeValue) {
        BRCryptoStatus status = LOOKUP.get(nativeValue);
        checkState(null != status);
        return status;
    }

    public abstract int toNative();
}

File: crypto/BRCryptoStatus.c
//
//  BRCryptoStatus.c
//  BRCore
//
//  Created by Michael Carrara on 7/31/19.
//  Copyright © 2019 breadwallet. All rights reserved.
//
//  See the LICENSE file at the project root for license information.
//  See the CONTRIBUTORS file at the project root for a list of contributors.

#include "BRCryptoStatusP.h"
#include "ethereum/BREthereum.h"

extern BRCryptoStatus
    cryptoStatusFromETH (BREthereumStatus status) {
    switch (status) {
        case SUCCESS:                           return CRYPTO_SUCCESS;

        // Reference access
        case ERROR_UNKNOWN_NODE:                return CRYPTO_ERROR_UNKNOWN_NODE;
        case ERROR_UNKNOWN_TRANSACTION:         return CRYPTO_ERROR_UNKNOWN_TRANSFER;
        case ERROR_UNKNOWN_ACCOUNT:             return CRYPTO_ERROR_UNKNOWN_ACCOUNT;
        case ERROR_UNKNOWN_WALLET:              return CRYPTO_ERROR_UNKNOWN_WALLET;
        case ERROR_UNKNOWN_BLOCK:               return CRYPTO_ERROR_UNKNOWN_BLOCK;
        case ERROR_UNKNOWN_LISTENER:            return CRYPTO_ERROR_UNKNOWN_LISTENER;

        // Node
        case ERROR_NODE_NOT_CONNECTED:          return CRYPTO_ERROR_NODE_NOT_CONNECTED;

        // Transfer
        case ERROR_TRANSACTION_HASH_MISMATCH:   return CRYPTO_ERROR_TRANSFER_HASH_MISMATCH;
        case ERROR_TRANSACTION_SUBMISSION:      return CRYPTO_ERROR_TRANSFER_SUBMISSION;

        // Number
        case ERROR_NUMERIC_PARSE:               return CRYPTO_ERROR_NUMERIC_PARSE;

        default:                                return CRYPTO_ERROR_FAILED;
    }
}

extern BREthereumStatus
    cryptoStatusAsETH (BRCryptoStatus status) {
    switch (status) {
        case CRYPTO_SUCCESS:                        return SUCCESS;

        // Reference access
        case CRYPTO_ERROR_UNKNOWN_NODE:             return ERROR_UNKNOWN_NODE;
        case CRYPTO_ERROR_UNKNOWN_TRANSFER:         return ERROR_UNKNOWN_TRANSACTION;
        case CRYPTO_ERROR_UNKNOWN_ACCOUNT:          return ERROR_UNKNOWN_ACCOUNT;
        case CRYPTO_ERROR_UNKNOWN_WALLET:           return ERROR_UNKNOWN_WALLET;
        case CRYPTO_ERROR_UNKNOWN_BLOCK:            return ERROR_UNKNOWN_BLOCK;
        case CRYPTO_ERROR_UNKNOWN_LISTENER:         return ERROR_UNKNOWN_LISTENER;

        // Node
        case CRYPTO_ERROR_NODE_NOT_CONNECTED:       return ERROR_NODE_NOT_CONNECTED;

        // Transfer
        case CRYPTO_ERROR_TRANSFER_HASH_MISMATCH:   return ERROR_TRANSACTION_HASH_MISMATCH;
        case CRYPTO_ERROR_TRANSFER_SUBMISSION:      return ERROR_TRANSACTION_SUBMISSION;

        // Number
        case CRYPTO_ERROR_NUMERIC_PARSE:            return ERROR_NUMERIC_PARSE;

        default:                                    return ERROR_FAILED;
    }
}
